keps/sig-node/3673-kubelet-parallel-image-pull-limit/README.md

```diff
@@ -0,0 +1,860 @@
+-->
+
+This KEP proposes adding to kubelet a node-level limit on the number of parallel image pulls.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### QPS/burst limits on kubelet are confusing
+Currently kubelet limits image pulls with QPS and burst, but they are confusing as they only limit the number of requests sent to container runtime, and does not consider the number of inflight image pulls. In other words, even a small QPS is set, there still could be many image pulls in progress in parallel, if each pull takes a long time. See [issue #112044](https://github.com/kubernetes/kubernetes/issues/112044) as an example.
+
+
+### No way to limit the number of inflight image pulls
+A user might want to put a limit on the number of images being pulled at the same time to avoid burden on disk IO or consuming too much network bandwidth. However, currently neither kubelet or containerd limits the number of inflight image pulls.
+
+On kubelet, as mentioned above, the QPS/burst limit does not take the in-progress pulls into account. On containerd, there is only [a per-image limit](https://github.com/containerd/containerd/blob/8e787543deede10f372b0e16ad5c07790fc680b6/pkg/cri/config/config.go#L299-L300), which only limits the number of parallel layer downloading for each image, but potentially allows unlimited number downloads overall.
+
+As a side node, cri-o has a configuration to limit the maximum layer downloads in-progress.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+Adding a node-level limit of parallel image pulls to kubelet. This limit will limit the maximum number of images being pulled in parallel. Any image pull request beyond the limit will be blocked until one image pull finishes.
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Prioritizing image pulls in any way.
+* Using the number of inflight image pulls as a signal to direct pod scheduling.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+* A new integer field `maxParallelImagePulls` will be added to Kubelet Configuration. `maxParallelImagePulls` is the maximum number of inflight image pulls.
```

Comment:
Not sure how we plan to incorporate Containerd's MaxConcurrentDownloads and CRI-O's configuration here to limit the total layers being pulled concurrently here.

Reply:
+1, interested in why this isn't left to the runtime, including its configuration.  Presumably other container runtime consumers have a similar problem.

Reply:
I think they are on a different scope. 
- image layer pull limit per image in container runtime: aim to less io overload when pulling an image with many layers.
an image can have 20 layers and if we pull them all together, that would be a big problem
- image pull limit per node in kubelet: aim to make a node not parel pull too many images
it is similar if only layers per image are limited. when we pull 20 images at the same time that is a problem.

Reply:
Yep Paco is correct, containerd's MaxConcurrentDownloads is the number of layers *per image*. To limit the total number of layers being downloaded, a user will need both maxParallelImagePulls and MaxConcurrentDownloads.

> why this isn't left to the runtime, including its configuration

Limiting the image pulls on kubelet can address the problem without runtime support, i.e. it will be useful regardless of the runtime used. It also kinda fixes https://github.com/kubernetes/kubernetes/issues/112044, by providing a real limit on kubernetes.

Reply:
I don't feel strongly about it, but it seems challenging for users to configure two different settings in kubelet and their runtime which ultimately try to control the same thing. The kubelet has less information (e.g. number of layers in the image), so it seems like the runtime can do a better job of determining what to do.  

e.g. If you had images with the same content, but different tags.  kubelet would artificially limit the number of pulls here while the runtime could do something smarter.


Sticking with the containerd example, suppose that `MaxConcurrentDownloads` worked across images in containerd.  Wouldn't that solve the problem better than an image pull limit on kubelet?  You could maximize your quota across images regardless of if they had few layers or many layers.

Reply:
https://github.com/containerd/containerd/issues/2195#issuecomment-372449427 I found an old thread about a similar debate here.

> If it is the latter, why can't we limit the concurrency in the upper system. For example, in Kubernetes we only use serial image download for production environment for now.

Reply:
BTW, I added some image pull-related metrics in containerd in https://github.com/containerd/containerd/pull/7313. 
```
	imagePulls           metrics.LabeledCounter
	inProgressImagePulls metrics.Gauge
	//  pull duration / (image size / 1MBi)
	imagePullThroughput prom.Histogram
```